compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp

```diff
@@ -874,7 +874,7 @@ decideFusableLinalgOps(Region &region, DominanceInfo const &dominanceInfo,
       // materializing large tensors between dispatches.
       if (!isa<linalg::LinalgOp, tensor::PadOp, tensor::PackOp,
                IREE::Encoding::SetEncodingOp>(op) ||
-          isa<linalg::FillOp>(op) || isBitExtendOp(&op)) {
+          isa<linalg::FillOp>(op)) {
```

Comment:
This change leaves the comment above outdated:
>       // Only look for Linalg ops here. Avoid moving `linalg.fill` that aren't
>       // fused with anything else into their own dispatches since it is better
>       // to convert them to splats. Also avoid moving dequantization-like ops
>       // into their own dispatch since it is better to clone these ops and avoid
>       // materializing large tensors between dispatches.

Can you elaborate on what specific issue you encountered?

Would a cost model with some heuristics work better than a coarse change like this affecting all programs?

Reply:
yes the specific issue I'm trying to mitigate is with this pattern (which is occurring in a few models we're trying to bring up from Onnx )
min repro:
```
module {
  func.func @largeVectorMinRepro(%arg0: tensor<1x320x65x65xi8>) -> tensor<1x320x1x1xf32> {
        %cst = arith.constant 1.250000e-01 : f32
        %cst_0 = arith.constant 0.000000e+00 : f32
        %c5408000 = arith.constant 5408000 : index
        %c0 = arith.constant 0 : index
        %3 = tensor.empty() : tensor<1x320x1x1xf32>
        %4 = tensor.empty() : tensor<65x65xf32>
        %5 = tensor.empty() : tensor<1x320x65x65xf32>
        %6 = linalg.fill ins(%cst_0 : f32) outs(%3 : tensor<1x320x1x1xf32>) -> tensor<1x320x1x1xf32>
        %7 = linalg.generic {indexing_maps = [affine_map<(d0, d1, d2, d3) -> (d0, d1, d2, d3)>, affine_map<(d0, d1, d2, d3) -> (d0, d1, d2, d3)>], iterator_types = ["parallel", "parallel", "parallel", "parallel"]} ins(%arg0 : tensor<1x320x65x65xi8>) outs(%5 : tensor<1x320x65x65xf32>) {
        ^bb0(%in: i8, %out: f32):
          %9 = arith.extsi %in : i8 to i32
          %10 = arith.sitofp %9 : i32 to f32
          %11 = arith.mulf %10, %cst : f32
          linalg.yield %11 : f32
        } -> tensor<1x320x65x65xf32>
        %8 = linalg.pooling_nchw_sum  ins(%7, %4 : tensor<1x320x65x65xf32>, tensor<65x65xf32>) outs(%6 : tensor<1x320x1x1xf32>) -> tensor<1x320x1x1xf32>
    return %8 : tensor<1x320x1x1xf32>
  }
}
```

compile command : `iree-compile --iree-input-demote-i64-to-i32 --iree-hal-target-backends=llvm-cpu largevectorissue.minrepro.mlir -o test.vmfb`

[host issue here](https://github.com/nod-ai/npu-benchmark/issues/3)

> Would a cost model with some heuristics work better than a coarse change like this affecting all programs?

I do understand this change affects all quantized models, 
hence I've just created this to get a discussion started on what would be the right way to do it and make progress on that :)

Reply:
I understand that it could fix all the quantized models because it is throwing quantization away in the IREE compiler. However, it is really not what we want.

I think it is codegen's responsibility to fix the issue. Avoiding `large vector sizes` in CPU backend should be easy because the tile sizes we used in pooling ops are large.

For this kind of issue, we can create an issue, attach the source artifacts, and ask codegen to fix it. If we identify that the dispatch formation can be improved, we go head to fix fusion logics or global-optimization logics.

RE how to dump source artifacts: you can pass `--iree-hal-dump-executable-sources-to=/tmp/dump` to `iree-compile`, and it will dump the source artifact (i.e., the mlir input for codegen) to `/tmp/dump`.

Reply:
Hey @hanhanW Thanks for your reply, it does clear somethings up

>  because it is throwing quantization away in the IREE compiler
This 👆 I'm not sure if I follow correctly. 
The way I understand this change works is, it will create standalone dispatches for all Q and DQ Ops and not eliminate them. This is also what I saw on compiling the minRepro shared in the comment above. 
So, I'm assuming that we are still doing Q/DQ just not within the same dispatch as their consumer. 

As I'm not familiar with the CPU codegen pipeline, I'm just seeking some clarity on what's really happening. would love to know your thoughts on the same

> For this kind of issue, we can create an issue, attach the source artifacts,...
I've created an IREE issue #18005 , it has the compile command + minRepro mlir (adapted from the actual failing dispatch). Since the model that we're working with is a private model I'm unable to share all the artefacts on a public issue but the this private issue should have more details. 
Please let me know what more information I can provide to help  move this forward.

Reply:
>> because it is throwing quantization away in the IREE compiler
>
> This 👆 I'm not sure if I follow correctly.
The way I understand this change works is, it will create standalone dispatches for all Q and DQ Ops and not eliminate them. This is also what I saw on compiling the minRepro shared in the comment above.
So, I'm assuming that we are still doing Q/DQ just not within the same dispatch as their consumer.

I think one of the quantization's goal is to have less memory allocation and data movement. Disabling the fusion results in larger buffer allocation which is a concern. So like with fusion, we could form `dispatch(i8_dq_op + matmul_i32i32i32)` and the input buffer has i8 type. Without fusion, it becomes `dispatch(i8_dq_op) + dispatch(matmul_i32i32i32)` and we allocate two additional i32 buffers for `LHS` and `RHS`. Hopefully, this answers your question. :)

Reply:
@PhaneeshB, "throwing quantization away" isn't referring to completely removing them from the program but rather materializing the high bit-width tensor. The reasoning behind quantization is that it decreases memory and having separate dispatches defeats this. For example, in the case of `Dequant -> Elementwise add 1.1 -> Quant` (where 'compute' is some linalg op, all ops _should_ be in the same dispatch. Any data that is transferred through dispatch boundaries must be written into memory, while ops in the same dispatch are implicitly fused (no transient allocations between ops). Here is some pseudocode to try to show it more clearly:

```
//=== 3 Dispatches ===//
vector<i8> input_tensor = ...
// Dispatch 1 (Dequantize): requires extra allocation of f32 tensor
vector<f32> temp_tensor = Dequant(input_tensor)
// Dispatch 2: Computation
for i in range(len(input_tensor)):
  tmp_tensor[i] = tmp_tensor[i] + 1.1
// Dispatch 3: quantize
input_tensor = Quant(tmp_tensor)

//=== 1 Dispatch (fused) ===//
tensor<i8> input_tensor = ...
for i in range(len(input_tensor)):
  // Q/DQ is preformed on each element
  f32 elem = Dequant(input_tensor[i]) + 1
  input_tensor[i] = Quant(elem)
```

So, keeping dequant/quant in the same dispatch (fusing them) ideally means that no "large" tensors get allocated and the high bit-width data never gets written to memory. I've encountered a similar issue https://github.com/iree-org/iree/issues/17642 but with a different cause. In both cases, the backend fails to find a way to bufferize the tensors and is forced to create stack allocations. It seems like the backend needs to get fixed to be more resilient to this.

I'm a bit surprised that this is the first time this issue has presented its self, maybe there is a flag somewhere that fixes this?

Reply:
@hanhanW and @IanWood1 Thank you so much guys, for taking the time out to explain this 🙏 
Now I have a better mental picture of the quantization puzzle.

>  I'm a bit surprised that this is the first time this issue has presented its self, maybe there is a flag somewhere that fixes this?

fwiw, Zach had pointed out that we could bypass the error by disabling failure check and increasing the stack size using these flags :
```
--iree-llvmcpu-fail-on-large-vector=0
--iree-llvmcpu-stack-allocation-limit=100000
```
But these are lower level llvm flags and only treating the symptom.

Reply:
Yes, it compile with the flags. But the generated code is very very very bad, so please do not rely on it. There are reasons to have the flags for development, but they are really not for other users.